docs/rfc/0696-p2p-node.adoc

```diff
@@ -0,0 +1,318 @@
+as defined below:
+
+[source,cddl]
+----
+request = [
+    request-headers,
+    ? payload: bstr,
+]
+response = ok / error
+ok = [
+    response-headers,
+    ? payload: bstr,
+]
+error = [
+    response-headers,
+    code: uint,
+    ? message: tstr,
+]
+request-headers = {
+    ua: client-id,
+    rq: request-id,
+    ? token: token,
+}
+response-headers = {
+    rq: request-id,
+}
+; Unambiguous, human-readable string identifying the client application. Mainly
+; for diagnostic purposes. Example: "radicle-link-cli/v1.2+deaf"
+client-id: tstr .size (4..16)
+; Request identifier, choosen by the client. The responder includes the
+; client-supplied value in the response, enabling request pipelining.
+;
+; Note that streaming / multi-valued responses may include the same id in
+; several response messages.
+request-id: bstr .size (4..16)
+
+; Placeholder for future one-time-token support.
+token: bstr
+----
+
+
+=== Request/Response payloads
+
+There is a supported set of requests a daemon will answer to which are as follows:
+
+[source,cddl]
+----
+request = discover / memebership / stats
+request-discover = {
+  type: "discover",
+  info: connect-info,
+}
+request-memebership = {
+  type: "membership",
+}
+request-stats = {
+  type: "stats",
+}
+
+response = membership / stats
+response-memebership = {
+  active: [peer-id],
+  passive: [peer-id],
+}
+response-stats = {
+  connections-total: uint,
+  connected-peers: [connect-info],
+  memebership-active: uint,
+  memebership-passive: uint,
+}
+
+connect-info = [
+  peer_id,
+  [address],
+]
+
+address = address-ipv4 / address-ipv6
+address-ipv4 = [
+  ip4,
+  port,
+]
+address-ipv6 = [
+  ip6,
+  port,
+]
+
+ip4 = bstr .size 4
+ip6 = bstr .size 16
+
+; Canonical representation of a peer.
+peer-id: bstr
+
+; Network port.
+port: uint
+----
+
+All types representing requests and responses and their serialisation logic MUST
+be exposed as linkable libraries. It is RECOMMENDED to also expose the
+functionality to communicate with the node via IPC as a library.
+
+== Operations
+
+=== Supervision
+
+Process supervision SHOULD be deferred to established system level service
+managers i.e. `<<systemd>>` and `<<launchd>>` for Linux and macOS respectively.
+To support both long-running as well as ad-hoc usage the daemon implementation
+SHALL be equipped with the ability to detect and read the information from its
+environment necessary to determine if it has been activated via socket. When
+binding to a socket it SHALL use the file descriptors provided by the init
+process. If none are provided it SHALL assume long-running operation and SHALL
+bind to the well-known path on a UNIX domain socket in mode `SOCK_STREAM` under
+Linux:
+
+  $XDG_RUNTIME_DIR/radicle/<srv>-<peer-id>.sock
+
+macOS:
+
+  TODO(xla): Define well-known path for darwin systems.
+
+If sufficiently focused library implementations are present in the ecosystem to
```

Comment:
Afaik, that'd be [`listenfd`](https://crates.io/crates/listenfd). It is unfortunately cheating on macOS and Win, so proper `launchd` registration would have to be added. The cheat for Win may actually be useful.

Reply:
Was eyeeing Mr. Ronachers libs as well for that purpose but actively refrained from making a recommendation in this document and leave it up to the implementation phase to address it. If you see a need to make this explicit before this document can be accepted I can expand with existing options that seem suitable.

Reply:
Then perhaps remove the sentence altogether -- what we need probably doesn't exist pret-a-porter, but if it did, it is obvious that we would try to collaborate.

Reply:
6964dee